examples/platform/linux/AppMain.cpp

```diff
@@ -120,15 +129,79 @@ int ChipLinuxAppInit(int argc, char ** argv)
     return 0;
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE
+
+using namespace ::chip;
+using namespace ::chip::Inet;
+using namespace ::chip::Transport;
+using namespace ::chip::DeviceLayer;
+using namespace ::chip::Messaging;
+using namespace ::chip::Controller;
+
+class MyServerStorageDelegate : public PersistentStorageDelegate
+{
+    CHIP_ERROR SyncGetKeyValue(const char * key, void * buffer, uint16_t & size) override
+    {
+        ChipLogProgress(AppServer, "Retrieved value from server storage.");
+        return PersistedStorage::KeyValueStoreMgr().Get(key, buffer, size);
+    }
+
+    CHIP_ERROR SyncSetKeyValue(const char * key, const void * value, uint16_t size) override
+    {
+        ChipLogProgress(AppServer, "Stored value in server storage");
+        return PersistedStorage::KeyValueStoreMgr().Put(key, value, size);
+    }
+
+    CHIP_ERROR SyncDeleteKeyValue(const char * key) override
+    {
+        ChipLogProgress(AppServer, "Delete value in server storage");
+        return PersistedStorage::KeyValueStoreMgr().Delete(key);
+    }
+};
+
+DeviceCommissioner gCommissioner;
+MyServerStorageDelegate gServerStorage;
+ExampleOperationalCredentialsIssuer gOpCredsIssuer;
+
+CHIP_ERROR InitCommissioner()
+{
+    NodeId localId = chip::kAnyNodeId;
+
+    chip::Controller::CommissionerInitParams params;
+
+    params.storageDelegate                = &gServerStorage;
+    params.mDeviceAddressUpdateDelegate   = nullptr;
+    params.operationalCredentialsDelegate = &gOpCredsIssuer;
+
+    ReturnErrorOnFailure(gOpCredsIssuer.Initialize(gServerStorage));
+
+    ReturnErrorOnFailure(gCommissioner.SetUdpListenPort(CHIP_PORT + 2));
```

Comment:
I assume this is being done because the client-like commissioner is being instantiated in a server-centric app.

I see why this would be necessary, as the controller/commissioner is doing this and cannot reuse the app's existing transport manager:

```
mTransportMgr          = chip::Platform::New<DeviceTransportMgr>();
```

It's a really big problem that the `chip::Controller::DeviceController` and `chip::Controller::DeviceCommissioner` objects insist on being the center of the universe.  This is definitely out of scope for this PR.  But this type of workaround should not be necessary.

Reply:
I think that type of cleanup will happen once we all get comfortable with both of these classes and evolve them to share more infrastructure. Right now this is not a problem because we are not trying to optimize for the kinds of devices that are both commissioners and commissionee's (eg. TVs).